### ISSUE TITLE
Problems parsing certificate with x509.decode_x509_certificate

### ISSUE THREAD

**Issue by luisza** [danni/python-pkcs11#8]
When I parse my DER-encoded certificate  with x509.decode_x509_certificate like this.

```
    from pkcs11.util import x509.decode_x509_certificate
    certs = []
    with token.open() as session:
         for cert in session.get_objects({
                 Attribute.CLASS: ObjectClass.CERTIFICATE}):
              certs.append(decode_x509_certificate(cert[Attribute.VALUE], extended_set=True))
```

I get this error.
![error_parsing_certificate](https://user-images.githubusercontent.com/7146904/29529381-adb02380-865c-11e7-8e09-b536255fb223.png)

My certificate is [here](https://github.com/luisza/dfva/blob/79906515e9aace19e2daf6be19c1dbc93deecdea/validator/tests/certificate.py#L1) , and is valid in Costa Rica also is ok if you parsing it with_PyOpenSSL_ or _asn1crypto_ .

**Comment by redNixon** [comment]
This will happen with any custom OID(or in this case a Microsoft one). Cryptography had a similar issue(https://github.com/pyca/cryptography/issues/2288) that ended with getting rid of pyasn1 due to its fragility.

I've dealt with a very similar issue before, moving most of util.x509 to use the cryptography library instead would probably be a good idea for stability, also a lot of util.x509's logic is also already implemented as well in cryptography.

I can work up a PR to do the migration if it is wanted or I can do a PR with some odd certificate configurations that I have found to be an issue in the past in other projects if you want to continue using pyans1 in order to smoke out the edgecases.

**Comment by danni** [comment]
I have a branch to start removing `pyasn1` in exchange for `asn1crypto` which can handle these certificates, and already includes a bunch of other code I'd needed to write. The biggest problem I'm not sure how to handle is that PKCS#11 defines the certificate `VALUE` needs to be BER encoded. But `asn1crypto` can only output `DER` encoding.

**Comment by danni** [comment]
https://github.com/danni/python-pkcs11/compare/asn1crypto-port

**Comment by redNixon** [comment]
Depending on how literal you want to take the standard, using DER should be fine since it is a more restrictive version of BER.  Personally I like the approach openSSL took with this: https://www.openssl.org/docs/faq.html#PROG4

asn1crypto will decode BER but will not encode it, so reading is not an issue but it would be for if python-pkcs11 stored an imported(?) value and something else accessing an HSM required literal BER later on

**Comment by danni** [comment]
Thanks!

**Comment by danni** [comment]
This is fixed in v0.4.0 with the port to `asn1crypto`.
